io/dagreader.go

```diff
@@ -66,12 +77,353 @@ func NewDagReader(ctx context.Context, n ipld.Node, serv ipld.NodeGetter) (DagRe
+// later move it to the output buffer (`Read`) or seek into it (`Seek`).
+func (dr *dagReader) saveNodeData(node ipld.Node) error {
+	extractedNodeData, err := unixfs.ReadUnixFSNodeData(node)
+	if err != nil {
+		return err
+	}
+
+	dr.currentNodeData = bytes.NewReader(extractedNodeData)
+	return nil
+}
+
+// Read the `currentNodeData` buffer into `out`. This function can't have
+// any errors as it's always reading from a `bytes.Reader` and asking only
+// the available data in it.
+func (dr *dagReader) readNodeDataBuffer(out []byte) int {
+
+	n, _ := dr.currentNodeData.Read(out)
+	// Ignore the error as the EOF may not be returned in the first
+	// `Read` call, explicitly ask for an empty buffer below to check
+	// if we've reached the end.
+
+	if dr.currentNodeData.Len() == 0 {
+		dr.currentNodeData = nil
+		// Signal that the buffer was consumed (for later `Read` calls).
+		// This shouldn't return an EOF error as it's just the end of a
+		// single node's data, not the entire DAG.
+	}
+
+	dr.offset += int64(n)
+	// TODO: Should `offset` be incremented here or in the calling function?
+	// (Doing it here saves LoC but may be confusing as it's more hidden).
+
+	return n
+}
+
+// WriteTo writes to the given writer.
+//
+// TODO: Improve performance. It would be better to progressively
+// write each node to the writer on every visit instead of allocating
+// a huge buffer, that would imply defining a `Visitor` very similar
+// to the one used in `CtxReadFull` (that would write to the `io.Writer`
+// instead of the reading into the `currentNodeData` buffer). More
+// consideration is needed to restructure those two `Visitor` functions
+// to avoid repeating code.
+func (dr *dagReader) WriteTo(w io.Writer) (int64, error) {
+	writeBuf, err := ioutil.ReadAll(dr)
+	if err != nil {
+		return 0, err
+	}
+	return bytes.NewReader(writeBuf).WriteTo(w)
+}
+
+// Close the reader (cancelling fetch node operations requested with
+// the internal context, that is, `Read` calls but not `CtxReadFull`
+// with user-supplied contexts).
+func (dr *dagReader) Close() error {
+	dr.cancel()
+	return nil
+}
+
+// Seek implements `io.Seeker` seeking to a given offset in the DAG file,
+// it matches the standard unix `seek`. It moves the position of the internal
+// `dagWalker` and may also leave a `currentNodeData` buffer loaded in case
+// the seek is performed to the middle of the data in a node.
+//
+// TODO: Support seeking from the current position (relative seek)
+// through the `dagWalker` in `io.SeekCurrent`.
+func (dr *dagReader) Seek(offset int64, whence int) (int64, error) {
+	switch whence {
+	case io.SeekStart:
+		if offset < 0 {
+			return -1, errors.New("invalid offset")
+		}
+
+		if offset == dr.offset {
+			return offset, nil
+			// Already at the requested `offset`, nothing to do.
+		}
+
+		left := offset
+		// Amount left to seek.
+
+		// Seek from the beginning of the DAG.
+		dr.resetPosition()
```

Comment:
Can we not do this if `offset > dr.offset`?

Reply:
I don't think so because of how our current file DAG system works. The check makes sense but we are normally on a leaf and currently the `Walker` can't go up (we should definitely improve on this in the `SeekCurrent` case, since at the moment we're lying to the user and seeking from the start). That is, the target may be "on the right" of our current position but we don't know how to get there unless we start from the root and follow its size hints.